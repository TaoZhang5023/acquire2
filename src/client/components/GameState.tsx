import { List } from 'immutable';
import * as React from 'react';

import { GameAction, GameBoardType } from '../../common/enums';
import { ActionBase } from '../../common/gameActions/base';
import { ActionDisposeOfShares } from '../../common/gameActions/disposeOfShares';
import { ActionGameOver } from '../../common/gameActions/gameOver';
import { ActionPlayTile } from '../../common/gameActions/playTile';
import { ActionPurchaseShares } from '../../common/gameActions/purchaseShares';
import { ActionSelectChainToDisposeOfNext } from '../../common/gameActions/selectChainToDisposeOfNext';
import { ActionSelectMergerSurvivor } from '../../common/gameActions/selectMergerSurvivor';
import { ActionSelectNewChain } from '../../common/gameActions/selectNewChain';
import { ActionStartGame } from '../../common/gameActions/startGame';
import { gameBoardTypeToCSSClassName, gameBoardTypeToHotelInitial, getHotelNameSpan, getUsernameSpan } from '../helpers';
import * as style from './GameState.css';

export interface GameStateProps {
    usernames: List<string>;
    nextGameAction: ActionBase;
    width: number;
    height: number;
}

export class GameState extends React.PureComponent<GameStateProps> {
    render() {
        const { usernames, nextGameAction, width, height } = this.props;

        return gameStateHandlerLookup[nextGameAction.gameAction](usernames, nextGameAction, width, height);
    }
}

// @ts-ignore
const gameStateHandlerLookup: { [key: number]: (usernames: List<string>, action: ActionBase, width: number, height: number) => JSX.Element } = {
    [GameAction.StartGame]: (usernames: List<string>, action: ActionStartGame, width: number, height: number) => (
        <div className={style.root} style={{ width, height }}>
            Waiting for {getUsernameSpan(usernames.get(action.playerID, ''))} to start the game.
        </div>
    ),
    [GameAction.PlayTile]: (usernames: List<string>, action: ActionPlayTile, width: number, height: number) => (
        <div className={style.root} style={{ width, height }}>
            Waiting for {getUsernameSpan(usernames.get(action.playerID, ''))} to play a tile.
        </div>
    ),
    [GameAction.SelectNewChain]: (usernames: List<string>, action: ActionSelectNewChain, width: number, height: number) => (
        <div className={style.root} style={{ width, height }}>
            Waiting for {getUsernameSpan(usernames.get(action.playerID, ''))} to select new chain ({getHotelInitialsList(action.availableChains)}).
        </div>
    ),
    [GameAction.SelectMergerSurvivor]: (usernames: List<string>, action: ActionSelectMergerSurvivor, width: number, height: number) => (
        <div className={style.root} style={{ width, height }}>
            Waiting for {getUsernameSpan(usernames.get(action.playerID, ''))} to select merger survivor ({getHotelInitialsList(action.chainsBySize[0])}).
        </div>
    ),
    [GameAction.SelectChainToDisposeOfNext]: (usernames: List<string>, action: ActionSelectChainToDisposeOfNext, width: number, height: number) => (
        <div className={style.root} style={{ width, height }}>
            Waiting for {getUsernameSpan(usernames.get(action.playerID, ''))} to select chain to dispose of next ({getHotelInitialsList(action.defunctChains)}).
        </div>
    ),
    [GameAction.DisposeOfShares]: (usernames: List<string>, action: ActionDisposeOfShares, width: number, height: number) => (
        <div className={style.root} style={{ width, height }}>
            Waiting for {getUsernameSpan(usernames.get(action.playerID, ''))} to dispose of {getHotelNameSpan(action.defunctChain)} shares.
        </div>
    ),
    [GameAction.PurchaseShares]: (usernames: List<string>, action: ActionPurchaseShares, width: number, height: number) => (
        <div className={style.root} style={{ width, height }}>
            Waiting for {getUsernameSpan(usernames.get(action.playerID, ''))} to purchase shares.
        </div>
    ),
    [GameAction.GameOver]: (usernames: List<string>, action: ActionGameOver, width: number, height: number) => (
        <div className={style.root} style={{ width, height }}>
            Game over.
        </div>
    ),
};

function getHotelInitialsList(chains: GameBoardType[]) {
    const entries: (JSX.Element | string)[] = new Array(chains.length * 2 - 1);

    for (let i = 0; i < chains.length; i++) {
        const chain = chains[i];
        entries[i * 2] = (
            <span key={chain} className={gameBoardTypeToCSSClassName[chain]}>
                {gameBoardTypeToHotelInitial[chain]}
            </span>
        );
    }

    if (chains.length === 2) {
        entries[1] = ' or ';
    } else {
        for (let i = 1; i < chains.length - 1; i++) {
            entries[i * 2 - 1] = ', ';
        }
        entries[(chains.length - 1) * 2 - 1] = ', or ';
    }

    return entries;
}
