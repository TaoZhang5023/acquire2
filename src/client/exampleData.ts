import { List } from 'immutable';
import { defaultMoveDataHistory } from '../common/defaults';
import { GameBoardType, GameHistoryMessage, GameMode, PlayerArrangementMode } from '../common/enums';
import { Game } from '../common/game';
import { ActionDisposeOfShares } from '../common/gameActions/disposeOfShares';
import { ActionGameOver } from '../common/gameActions/gameOver';
import { ActionPlayTile } from '../common/gameActions/playTile';
import { ActionPurchaseShares } from '../common/gameActions/purchaseShares';
import { ActionSelectChainToDisposeOfNext } from '../common/gameActions/selectChainToDisposeOfNext';
import { ActionSelectMergerSurvivor } from '../common/gameActions/selectMergerSurvivor';
import { ActionSelectNewChain } from '../common/gameActions/selectNewChain';
import { ActionStartGame } from '../common/gameActions/startGame';
import { getNewTileBag } from '../common/helpers';
import { allChains } from './helpers';

export function getDummyGameForGetGameHistory() {
    const game = new Game(GameMode.Singles4, PlayerArrangementMode.ExactOrder, getNewTileBag(), List([2, 3, 5, 8]), List(['Tim', 'Rita', 'Dad', 'Mom']), 8, 3);
    game.doGameAction([], null);
    game.moveDataHistory = defaultMoveDataHistory;

    let moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.TurnBegan, 0, []);
    moveData.timestamp = 1524896229792;
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.DrewPositionTile, 1, [21]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.StartedGame, 2, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.DrewTile, 3, [100]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.HasNoPlayableTile, 0, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.PlayedTile, 1, [40]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.FormedChain, 2, [0]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.MergedChains, 3, [[1, 2]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.MergedChains, 0, [[3, 4, 5]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.MergedChains, 1, [[0, 1, 2, 6]]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.SelectedMergerSurvivor, 2, [3]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.SelectedChainToDisposeOfNext, 3, [4]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.ReceivedBonus, 0, [5, 25]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.DisposedOfShares, 1, [6, 2, 3]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.CouldNotAffordAnyShares, 2, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.PurchasedShares, 3, [[]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.PurchasedShares, 0, [[[0, 3]]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.PurchasedShares, 1, [[[1, 2], [2, 1]]]);
    moveData.addGameHistoryMessage(GameHistoryMessage.PurchasedShares, 2, [[[3, 1], [4, 1], [5, 1]]]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.DrewLastTile, 3, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.ReplacedDeadTile, 0, [30]);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.EndedGame, 1, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.NoTilesPlayedForEntireRound, null, []);
    game.endCurrentMove();

    moveData = game.getCurrentMoveData();
    moveData.addGameHistoryMessage(GameHistoryMessage.AllTilesPlayed, null, []);
    game.endCurrentMove();

    return game;
}

export function getExampleNextGameActionsArray() {
    const game = new Game(
        GameMode.Singles5,
        PlayerArrangementMode.RandomOrder,
        [],
        List([1, 2, 3, 4, 5]),
        List(['Tim', 'Rita', 'Dad', 'Mom', 'REALLY, REALLY, REALLY, REALLY, REALLY LONG NAME']),
        1,
        6,
    );
    return [
        new ActionStartGame(game, 4),
        new ActionStartGame(game, 0),
        new ActionPlayTile(game, 1),
        new ActionSelectNewChain(game, 2, allChains, 107),
        new ActionSelectMergerSurvivor(game, 3, [GameBoardType.Luxor, GameBoardType.Festival, GameBoardType.Continental], 107),
        new ActionSelectChainToDisposeOfNext(game, 0, [GameBoardType.Tower, GameBoardType.American], GameBoardType.Continental),
        new ActionDisposeOfShares(game, 1, GameBoardType.Imperial, GameBoardType.Luxor),
        new ActionPurchaseShares(game, 2),
        new ActionGameOver(game, 3),
    ];
}
