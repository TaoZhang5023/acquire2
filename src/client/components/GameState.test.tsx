import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { getExampleNextGameActionsArray, getUsernamesForExampleNextGameActionsArray } from '../exampleData';
import { GameState } from './GameState';

describe('GameState', () => {
    it('renders correctly', () => {
        const gameStateUsernames = getUsernamesForExampleNextGameActionsArray();
        const nextGameActionsArray = getExampleNextGameActionsArray();

        const component = renderer.create(
            <div>
                {nextGameActionsArray.map((nextGameAction, i) => (
                    <GameState key={i} usernames={gameStateUsernames} nextGameAction={nextGameAction} width={500} height={22} />
                ))}
            </div>,
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
