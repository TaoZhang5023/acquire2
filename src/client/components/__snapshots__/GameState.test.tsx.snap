// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameState renders correctly 1`] = `
<div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Waiting for 
    <span
      className="username"
    >
      REALLY, REALLY, REALLY LONG NAME
    </span>
     to start the game.
  </div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Waiting for 
    <span
      className="username"
    >
      Tim
    </span>
     to start the game.
  </div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Waiting for 
    <span
      className="username"
    >
      Rita
    </span>
     to play a tile.
  </div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Waiting for 
    <span
      className="username"
    >
      Dad
    </span>
     to select new chain (
    <span
      className="luxor"
    >
      L
    </span>
    , 
    <span
      className="tower"
    >
      T
    </span>
    , 
    <span
      className="american"
    >
      A
    </span>
    , 
    <span
      className="festival"
    >
      F
    </span>
    , 
    <span
      className="worldwide"
    >
      W
    </span>
    , 
    <span
      className="continental"
    >
      C
    </span>
    , or 
    <span
      className="imperial"
    >
      I
    </span>
    ).
  </div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Waiting for 
    <span
      className="username"
    >
      Mom
    </span>
     to select merger survivor (
    <span
      className="luxor"
    >
      L
    </span>
    , 
    <span
      className="festival"
    >
      F
    </span>
    , or 
    <span
      className="continental"
    >
      C
    </span>
    ).
  </div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Waiting for 
    <span
      className="username"
    >
      Tim
    </span>
     to select chain to dispose of next (
    <span
      className="tower"
    >
      T
    </span>
     or 
    <span
      className="american"
    >
      A
    </span>
    ).
  </div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Waiting for 
    <span
      className="username"
    >
      Rita
    </span>
     to dispose of 
    <span
      className="imperial"
    >
      Imperial
    </span>
     shares.
  </div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Waiting for 
    <span
      className="username"
    >
      Dad
    </span>
     to purchase shares.
  </div>
  <div
    className="root"
    style={
      Object {
        "height": 22,
        "width": 500,
      }
    }
  >
    Game over.
  </div>
</div>
`;
